### ISSUE TITLE
lshift/rshift bug

### ISSUE THREAD

**Issue by ehoffman2** [bitcoin-sv/bitcoin-sv#50]
Hello

There seem to be a bug in the RShift and LShift functions, in src/script/interpreter.cpp (multiple issues actually)

I'm trying to figure out exactly the logic of it, but normally, doing a right-shift by n basically is equivalent to a divide by 2^n.  And similarly, doing a lshift by n is equivalent to multiply by 2^n.

All the numbers in Bitcoins are treated as little endian, this is true for CScriptNum (see set_vch() and getvch()), and also for the hashes (for example, hash256 are displayed as if the 32-byte array is a single little-endian big number, with the LSB at index 0, MSB at index 31).

So, doing a LShift/RShift should follow the same endianness, meaning that doing a RShift by 1 would take an array and do the following:

result = RShift(x,1):
- x[0], bits [0] ==> (dropped)
- x[0], bits [7:1] ==> result[0] bits [6:0]
- x[1], bits [0] ==> result[0] bits [7]
- x[1], bits [7:1] ==> result[1] bits [6:0]
- x[2], bits [0] ==> result[1] bits [7]

But, looking at the RShift function (LShift have the same issue)

```
inline uint8_t make_rshift_mask(size_t n) {
    static uint8_t mask[] = {0xFF, 0xFE, 0xFC, 0xF8, 0xF0, 0xE0, 0xC0, 0x80}; 
    return mask[n]; 
} 

static valtype RShift(const valtype &x, int n) {
    int bit_shift = n % 8; 
    int byte_shift = n / 8; 
 
    uint8_t mask = make_rshift_mask(bit_shift); 
    uint8_t overflow_mask = ~mask; 
 
    valtype result(x.size(), 0x00); 
    for (int i = 0; i < (int)x.size(); i++) {
        int k = i + byte_shift;
        if (k < (int)x.size()) {
            uint8_t val = (x[i] & mask); 
            val >>= bit_shift;
            result[k] |= val; 
        } 

        if (k + 1 < (int)x.size()) {
            uint8_t carryval = (x[i] & overflow_mask); 
            carryval <<= 8 - bit_shift; 
            result[k + 1] |= carryval;
        } 
    } 
    return result; 
} 

```
There are multiple issues here.

First, i is used as a source index, and k as a destination index.  And all the operations involve moving data from x[i] (which iterate from 0 to x.size()-1), to result[k] (which k is set to i+byte_shift).  So, you basically do a 'multiply by 2^n' instead of 'divide by 2^n'.

Second issue, let's assume that RShift and LShift logic is reversed, there is still a flaw in shift logic.  If you take RShift again, and you look at the first bit mask/shift block, you do see a right-shift operator, meaning that the bytes themselves are really right-shifted (divided by 2).

Ex: if you take RShift(x,1), then x[0] bits [7:1] --> result[0] bits [6:0].

So, you have an identity crisis here, specifically: Is the RShift a 'multiply by 2^n' or 'divide by 2^n'?!

3rd issue, if you look at the second bit mask/shift block, it does not end up in a logical arrangement.  Let's say you have a value with N bytes.  When you do RShift, by 1 bit for example, you would expect that the byte 0 would have it's bit shifted right by 1:
- x[0], bits [0] ==> (dropped)
- x[0], bits [7:1] ==> result[0] bits [6:0]
- x[1], bits [0] ==> result[0] bits [7]
- x[1], bits [7:1] ==> result[1] bits [6:0]
- x[2], bits [0] ==> result[1] bits [7]
and so on...

However, the carry is took from the wrong index (i.e. x[i] instead of x[i+1]) and carried over to the wrong destination index (i.e. result[k+1] instead of filling in the remaining bits of result[k] that was partially filled in the operation just above).  The above will do:

- x[0], bits [7:1] ==> result[0] bits [6:0]
- x[0], bits [0] ==> result[1] bits [7]
- x[1], bits [7:1] ==> result[1] bits [6:0]
- x[1], bits [0] ==> result[2] bits [7]
and so on...

Note that the second block would need to adjust the bound check accordingly.

So, basically, RShift and LShift does not work in an expected way, and in any logical way.

I would need second confirmation though...  What do you think?

Regards,
Eric

**Comment by kyuupichan** [comment]
The shift operations treat the value as a bitstring.  Since Bitcoin originally used SSL it acquired a little-endian representation which means you cannot reconcile arithermetic and bitstring interpretation of shifts simultaneously.  Bitstring was chosen.

**Comment by ehoffman2** [comment]
So, the way this work right now is that the bit string is represented as:

```
      +---------------+---------------+---------------+-
Byte  |       0       |       1       |       2       | ...
      +---------------+---------------+---------------+-
Bit   |7 6 5 4 3 2 1 0|7 6 5 4 3 2 1 0|7 6 5 4 3 2 1 0| ...
      +---------------+---------------+---------------+-

                     RShift -------->
                     <-------- LShift

```
Ok, still look odd, but I gave more taught to it.  And, in the end, if you look at it, SHA functions does work in this way, along with many serialized protocols where the most significant bit is sent first.  For example, for many crypto functions, like SHA-256, at some point you have to add a '1' at the end of the 'stream', before doing the crypto rounds.  With byte-oriented computer software, you often see a 0x80 added (which really is a '1' bit added to a multiple of 8-bit bytes).

I still do not see a case where this functionality would be of use on the blockchain, other than a curiosity, but that's another debate :-)

This should be somehow documented in a wiki...

Thanks,
Eric

**Comment by Danconnolly** [comment]
Closing.
